### ISSUE TITLE
Make Go To Definition modifier key configurable

### ISSUE THREAD

**Issue by ChristopherHaws** [Microsoft/VS-PPT#9]
I really like the idea of the mouse clickable "Go to Definition", however I use Ctrl+Click to select a whole word so often that I always have to disable this feature. I would like to make this configurable so that I could make the modifier key into the Alt key. I would be more than happy to add this feature myself if you approve it. :)

Thanks for open sourcing this great tool!

**Comment by justcla** [comment]
Hi Christopher. Thanks for your suggestion.

I suppose what you suggest is possible. And there should be no harm in approving a configurable modifier key. But what would be the interface for the user to set the modifier key?

Before we do that, though, my first suggestion would be that you try to break the habit of using Ctrl-Click to select a word. I know it's out of the box behavior in VS, but Ctrl-Click to follow a link is much more common outside of VS, so I like using Ctrl-Click inside VS to GoToDefinition (personally), and I tend to use Double-Click to select the whole word.

Regardless, now that we have open-sourced the code, you can make this change locally - and use it locally. You won't need any approval from us to do that. You could even hard-code the key in that case.
(Just make sure you disable the original Ctrl-Click GoToDefinition in PPT before you install the replacement standalone extension! Else you might get unpredictable behavior.)

**Comment by ChristopherHaws** [comment]
I was thinking that a dropdown could be added to the "Other Extensions" options page in the options dialog to specify the modifier key. The default value could continue to be the control key, so it wouldn't change the default behavior.

I use ctrl+click for more than selecting one word, so double clicking wouldn't work for me. I tend to ctrl+click the first word and ctrl+shift+click a second word to select multiple words. I admit, this is a habitual action, but I am able to make selections much faster than the alternative.

**Comment by olegtk** [comment]
Good point about a conflict with Ctrl+click selecting a word though... I think we have to fix this by allowing a choice of other modifier keys as you suggested.

**Comment by mserfass** [comment]
I use Resharper heavily. I like and use the R# rich mouse navigation (ctrl+left click goes to declaration, ctrl+alt+left click goes to implementation). Unfortunately, R# doesn't do peek definition. It would be great to be able to use this add-in to invoke peek definition on ctrl+shift+left click, for example.

Could you add functionality for doing both peek definition on one user-defined mouse event and go to definition on a second user-defined mouse action?

As for the setting ui, what about check boxes for ctrl, alt, shift, and a radio for mouse button left/right/middle? Or do it like the VS settings keyboard mapping, where you press the keys you want to map to?

Thanks or making this add-in, and the others in the Power Tools collection. They're used and appreciated.

**Comment by justcla** [comment]
Previous comment illuded to issue #114.

Nice suggestions about the UI. Instead of checkboxes for the modifier keys, I suggest a drop-down with the following options:
- Ctrl
- Alt
- Shift
- Ctrl+Alt
- Ctrl+Shift
- Alt+Shift
- Ctrl+Alt+Shift
Three checkboxes for Ctrl, Alt and Shift would probably work equally well.

**Comment by justcla** [comment]
@mserfass I wanted to let you know that your ideas have spawned rigorous discussions around the Editor team. We like your idea of providing users the ability to easily choose GoToDefn or PeekToDefn using a modifier key.
@ChristopherHaws
We are also looking to provide users with the ability to choose their own GoToDefn modifier keys.

Our current thoughts are to match ReSharper's current choices for GoToDefn (Ctrl+Click) and PeekToDefn (Ctrl+Shift+Click).
And we will provide users the ability to choose the modifier for GoToDefn from Ctrl, Alt, or Ctrl+Alt.
Adding Shift to the chosen GoToDefn keys will change it to a PeekToDefn call.
(So if a user chooses Alt as the GoToDefn key, then Shift+Alt will fire off PeekToDefn.)

What do you think of that?

We currently have no implementation of Go To Implementation.
If and when we do, we will need to reconsider our options.
We are tentatively reserving Ctrl+Alt+Click for GoToImpl.

We haven't considered offering a choice of mouse buttons.
Middle-Click GoToDefn seems like a sensible default.

**Comment by justcla** [comment]
Note: There is a user voice about [this item](https://visualstudio.uservoice.com/forums/121579-visual-studio-ide/suggestions/2660830-add-ctrl-click-go-to-definition-from-productivity).

"I’m pleased to say we are looking to include this feature in a future release of Visual Studio."

As such, these changes we are discussing, will probably not be implemented in PPT, but instead, will appear in the core product.

@mserfass Regarding the modifier for PeekToDefn:
If you open a User Voice item about the modifier for PeekToDefn, we can track votes and keep you updated on our progress as we consider building that into the product too.

Thanks for your suggestions.

**Comment by mserfass** [comment]
Hi!

Sorry for the slow response. I had a tight deadline on a hardware and OS
combined change.

I like those ideas. I think it would please most of the users and fit the
most use cases. Keeping the key combos in sync between defn and peek keeps
it simple and easy to remember (and easier for you to code). Great for us
older devs who are getting senile. 😀

I'll open a user voice for votes.

Now I need a way to map the other six mouse buttons directly in VS. (Yes, I
use a gaming mouse when I develop.)


On Apr 17, 2017 20:46, "Justin Clareburt" <notifications@github.com> wrote:

@mserfass <https://github.com/mserfass> I wanted to let you know that your
ideas have spawned rigorous discussions around the Editor team. We like
your idea of providing users the ability to easily choose GoToDefn or
PeekToDefn using a modifier key.
@ChristopherHaws <https://github.com/ChristopherHaws>
We are also looking to provide users with the ability to choose their own
GoToDefn modifier keys.

Our current thoughts are to match ReSharper's current choices for GoToDefn
(Ctrl+Click) and PeekToDefn (Ctrl+Shift+Click).
And we will provide users the ability to choose the modifier for GoToDefn
from Ctrl, Alt, or Ctrl+Alt.
Adding Shift to the chosen GoToDefn keys will change it to a PeekToDefn
call.
(So if a user chooses Alt as the GoToDefn key, then Shift+Alt will fire off
PeekToDefn.)

What do you think of that?

We currently have no implementation of Go To Implementation.
If and when we do, we will need to reconsider our options.
We are tentatively reserving Ctrl+Alt+Click for GoToImpl.

We haven't considered offering a choice of mouse buttons.
Middle-Click GoToDefn seems like a sensible default.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/VS-PPT/issues/9#issuecomment-294638479>, or mute
the thread
